### PR TITLE
Add hardware model support

### PR DESCRIPTION
**Pull request by reuvenperetz**

### CODE REVIEW

File: docs/_sources/api/api_docs/modules/hardware_representation.rst.txt [hunk 1]
```diff
@@ -0,0 +1,23 @@
+:orphan:
+
+.. _ug-hardware_representation:
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply docs/_sources/api/api_docs/modules/hardware_representation.rst.txt]: right. fixed

File: docsrc/source/api/api_docs/modules/hardware_representation.rst [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+=================================
+hardware_representation Module
+=================================
+
+MCT can be configured to quantize and optimize models for different hardware settings.
+For example, when using qnnpack backend for Pytorch model inference, Pytorch quantization configuration
+uses per-tensor weights quantization for Conv2d, while when using tflite modeling, Tensorflow uses
+per-channel weights quantization for Conv2D.
+This can be addressed in MCT by using the hardware_representation module, that can configure different
```
Comment (haihabi) [inline docsrc/source/api/api_docs/modules/hardware_representation.rst:14]: Please put links to the models and make sure that the per-tensor and channel is correct

File: model_compression_toolkit/common/hardware_representation/hardware_model_component.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from model_compression_toolkit.common.hardware_representation.current_hardware_model import _current_hardware_model
+
+
+class HardwareModelComponent:
```
Comment (haihabi) [inline model_compression_toolkit/common/hardware_representation/hardware_model_component.py:19]: Missing comments

File: model_compression_toolkit/common/hardware_representation/op_quantization_config.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import copy
+from enum import Enum
+from typing import List
+
+class QuantizationMethod(Enum):
```
Comment (haihabi) [inline model_compression_toolkit/common/hardware_representation/op_quantization_config.py:20]: There are two QuantizationMethod enums

File: model_compression_toolkit/common/hardware_representation/operators.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+from model_compression_toolkit.common.hardware_representation.current_hardware_model import _current_hardware_model
+from model_compression_toolkit.common.hardware_representation.op_quantization_config import QuantizationConfigOptions
+
+
+class OperatorsSetBase(HardwareModelComponent):
+    """
+    Base class to represent a set of operators.
+    """
+    def __init__(self, name: str):
+        super().__init__(name=name)
```
Comment (haihabi) [inline model_compression_toolkit/common/hardware_representation/operators.py:27]: Comments

File: model_compression_toolkit/common/immutable.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    Class to make inherits immutable
+    """
+
+    _initialized = False
+
+    def __init__(self):
+        self._initialized = False
+
+    def __setattr__(self, *args, **kwargs):
+        if self._initialized:
```
Comment (haihabi) [inline model_compression_toolkit/common/immutable.py:27]: Comments

File: model_compression_toolkit/common/post_training_quantization.py [hunk 7]
```diff
@@ -459,8 +462,7 @@ def _prepare_model_for_quantization(in_model: Any,
     # Shift Negative Activations
     ######################################
     if quant_config.shift_negative_activation_correction and \
-            quant_config.enable_activation_quantization and \
-            quant_config.activation_quantization_method is not QuantizationMethod.UNIFORM:
+            quant_config.enable_activation_quantization:
```
Comment (haihabi) [inline model_compression_toolkit/common/post_training_quantization.py:465]: Why has the condition here changed?

File: docsrc/source/api/api_docs/modules/hardware_representation.rst [hunk 8]
```diff
@@ -0,0 +1,340 @@
+:orphan:
+
+.. _ug-hardware_representation:
+
+
+=================================
+hardware_representation Module
+=================================
```
Comment (haihabi) [inline docsrc/source/api/api_docs/modules/hardware_representation.rst:8]: Please add a note the current we only support Quantizer type in HM

File: model_compression_toolkit/common/quantization/set_node_quantization_config.py [hunk 9]
```diff
@@ -97,21 +106,28 @@ def create_node_activation_qc(qc: QuantizationConfig,
         Activation quantization configuration of a node.
     """
 
-    activation_quantization_fn = fw_info.activation_quantizer_mapping.get(qc.activation_quantization_method)
+    # Temporal assertion as this is the only use for hw model right now
+    for cfg in op_cfg_options.quantization_config_list:
```
Comment (haihabi) [inline model_compression_toolkit/common/quantization/set_node_quantization_config.py:110]: Why do we need this assertion?

File: model_compression_toolkit/common/quantization/set_node_quantization_config.py [hunk 10]
```diff
@@ -153,15 +171,25 @@ def _create_node_candidates_weights_qc(qc: QuantizationConfig,
     Returns:
         List of candidates of weights quantization configurations to set for a node.
     """
+    # Temporal assertion as this is the only use for hw model right now
+    for cfg in op_cfg_options.quantization_config_list:
+        assert cfg.weights_quantization_method == op_cfg_options.quantization_config_list[0].weights_quantization_method
+    op_cfg = op_cfg_options.quantization_config_list[0]
```
Comment (haihabi) [inline model_compression_toolkit/common/quantization/set_node_quantization_config.py:177]: Same as above

File: model_compression_toolkit/hardware_models/default_hwm.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+                             enable_activation_quantization=False))
+
+        # To quantize a model using mixed-precision, create
+        # a QuantizationConfigOptions with more than one QuantizationConfig.
+        # In this example, we quantize some operations' weights
+        # using 2, 4 or 8 bits, and when using 2 or 4 bits, it's possible
+        # to quantize the operations' activations using LUT.
+        four_bits = eight_bits.clone_and_edit(weights_n_bits=4)
+        two_bits = eight_bits.clone_and_edit(weights_n_bits=2)
+        four_bits_lut = four_bits.clone_and_edit(
```
Comment (haihabi) [inline model_compression_toolkit/hardware_models/default_hwm.py:67]: Remove the LUT only weights quantization and 8,4,2 MP

File: model_compression_toolkit/hardware_models/get_model.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    from model_compression_toolkit.hardware_models.keras_hardware_model.keras_tflite import KERAS_TFLITE_MODEL
+
+    tf_models_dict = {'default': KERAS_DEFAULT_MODEL,
+                      'tflite': KERAS_TFLITE_MODEL}
+if found_torch:
+    from model_compression_toolkit.hardware_models.pytorch_hardware_model.pytorch_default import PYTORCH_DEFAULT_MODEL
+    from model_compression_toolkit.hardware_models.pytorch_hardware_model.pytorch_qnnpack import PYTORCH_QNNPACK_MODEL
+
+    torch_models_dict = {'default': PYTORCH_DEFAULT_MODEL,
+                         'qnnpack': PYTORCH_QNNPACK_MODEL}
```
Comment (haihabi) [inline model_compression_toolkit/hardware_models/get_model.py:38]: Add pytorch tflite and qnnpack tensorflow

File: model_compression_toolkit/hardware_models/keras_hardware_model/keras_default.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        hwm.OperationsSetToLayers("Sigmoid", [tf.nn.sigmoid,
+                                              hwm.LayerFilterParams(Activation, activation="sigmoid")])
+
+        hwm.OperationsSetToLayers("Tanh", [tf.nn.tanh,
+                                           hwm.LayerFilterParams(Activation, activation="tanh")])
+
+    return default_hwm_keras
+
+
+KERAS_DEFAULT_MODEL = get_default_hwm_keras()
```
Comment (haihabi) [inline model_compression_toolkit/hardware_models/keras_hardware_model/keras_default.py:73]: Why do we need this constant? relevant to all models

File: model_compression_toolkit/keras/quantization_facade.py [hunk 14]
```diff
@@ -29,9 +29,12 @@
 
 import importlib
 
+from model_compression_toolkit.common.hardware_representation.hardware2framework import FrameworkHardwareModel
+
 if importlib.util.find_spec("tensorflow") is not None\
         and importlib.util.find_spec("tensorflow_model_optimization") is not None:
     import tensorflow as tf
+    from model_compression_toolkit.hardware_models.keras_hardware_model.keras_default import KERAS_DEFAULT_MODEL
```
Comment (haihabi) [inline model_compression_toolkit/keras/quantization_facade.py:37]: Please define this  KERAS_DEFAULT_MODEL where using get_model function

File: model_compression_toolkit/pytorch/quantization_facade.py [hunk 15]
```diff
@@ -31,17 +32,18 @@
 if importlib.util.find_spec("torch") is not None:
     from model_compression_toolkit.pytorch.default_framework_info import DEFAULT_PYTORCH_INFO
     from model_compression_toolkit.pytorch.pytorch_implementation import PytorchImplementation
+    from model_compression_toolkit.hardware_models.pytorch_hardware_model.pytorch_default import PYTORCH_DEFAULT_MODEL
```
Comment (haihabi) [inline model_compression_toolkit/pytorch/quantization_facade.py:35]: As a TensorFlow use the get of the DEFAULT_MODEL

File: tests/keras_tests/feature_networks_tests/base_keras_feature_test.py [hunk 16]
```diff
@@ -37,7 +37,8 @@ def __init__(self,
                          num_of_inputs=num_of_inputs,
                          input_shape=input_shape)
 
-
+    def get_fw_hw_model(self):
+        return KERAS_DEFAULT_MODEL
```
Comment (haihabi) [inline tests/keras_tests/feature_networks_tests/base_keras_feature_test.py:41]: use get function

File: model_compression_toolkit/common/hardware_representation/op_quantization_config.py (hunk 4 above)
Comment (reuvenperetz) [reply model_compression_toolkit/common/hardware_representation/op_quantization_config.py:20]: No... I moved it from QuantizationConfig. You think it should stay there?

File: model_compression_toolkit/common/hardware_representation/operators.py (hunk 5 above)
Comment (reuvenperetz) [reply model_compression_toolkit/common/hardware_representation/operators.py:32]: Fixed

File: model_compression_toolkit/common/immutable.py (hunk 6 above)
Comment (reuvenperetz) [reply model_compression_toolkit/common/immutable.py:38]: Fixed

File: model_compression_toolkit/common/post_training_quantization.py (hunk 7 above)
Comment (reuvenperetz) [reply model_compression_toolkit/common/post_training_quantization.py:465]: Cause quant_config does not have activation_quantization_method  anymore.

File: docsrc/source/api/api_docs/modules/hardware_representation.rst (hunk 8 above)
Comment (reuvenperetz) [reply docsrc/source/api/api_docs/modules/hardware_representation.rst:8]: Fixed

File: model_compression_toolkit/common/quantization/set_node_quantization_config.py (hunk 9 above)
Comment (reuvenperetz) [reply model_compression_toolkit/common/quantization/set_node_quantization_config.py]: Fixed

File: model_compression_toolkit/hardware_models/default_hwm.py (hunk 11 above)
Comment (reuvenperetz) [reply model_compression_toolkit/hardware_models/default_hwm.py]: Fixed

File: model_compression_toolkit/hardware_models/get_model.py (hunk 12 above)
Comment (reuvenperetz) [reply model_compression_toolkit/hardware_models/get_model.py]: In progress...

File: model_compression_toolkit/hardware_models/keras_hardware_model/keras_default.py (hunk 13 above)
Comment (reuvenperetz) [reply model_compression_toolkit/hardware_models/keras_hardware_model/keras_default.py]: fixed

File: model_compression_toolkit/keras/quantization_facade.py (hunk 14 above)
Comment (reuvenperetz) [reply model_compression_toolkit/keras/quantization_facade.py]: fixed

File: model_compression_toolkit/pytorch/quantization_facade.py (hunk 15 above)
Comment (reuvenperetz) [reply model_compression_toolkit/pytorch/quantization_facade.py]: fixed

File: docsrc/source/api/api_docs/modules/hardware_representation.rst [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+==========================
+Select a method to use during quantization:
+
+.. autoclass:: model_compression_toolkit.hardware_representation.QuantizationMethod
+
+
+|
+
+
+.. note::
```
Comment (haihabi) [inline docsrc/source/api/api_docs/modules/hardware_representation.rst:36]: Please put it in the beginning of the page and in the get function

File: model_compression_toolkit/pytorch/quantization_facade.py [hunk 18]
```diff
@@ -33,6 +34,8 @@
     from model_compression_toolkit.pytorch.pytorch_implementation import PytorchImplementation
     from torch.nn import Module
 
+    from model_compression_toolkit import get_model
+    PYTORCH_DEFAULT_MODEL = get_model('pytorch', 'default')
```
Comment (haihabi) [inline model_compression_toolkit/pytorch/quantization_facade.py:38]: Make the string constants

File: model_compression_toolkit/keras/quantization_facade.py [hunk 19]
```diff
@@ -39,6 +41,9 @@
     from model_compression_toolkit.keras.gradient_ptq.gptq_loss import multiple_tensors_mse_loss
     from keras.optimizer_v2.optimizer_v2 import OptimizerV2
 
+    from model_compression_toolkit import get_model
+    KERAS_DEFAULT_MODEL = get_model('tensorflow', 'default')
```
Comment (haihabi) [inline model_compression_toolkit/keras/quantization_facade.py:45]: Make the string constants

File: model_compression_toolkit/common/post_training_quantization.py (hunk 7 above)
Comment (haihabi) [reply model_compression_toolkit/common/post_training_quantization.py:465]: Still we need this check let talk about it

File: model_compression_toolkit/common/hardware_representation/op_quantization_config.py (hunk 4 above)
Comment (haihabi) [reply model_compression_toolkit/common/hardware_representation/op_quantization_config.py:20]: Please check again there is the same enum twice.
Line 20 and 201

**Pull request merged**